### PR TITLE
glfw

### PR DESCRIPTION
**Pull request by uhuntu**
This is my first try to implement by glfw

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,17 +7,18 @@ build = "build.rs"
 description = "A simple GUI for Neovim."
 
 [features]
-default = ["sdl2"]
+default = ["glfw"]
```
Comment (j4qfrost) [inline Cargo.toml:10]: Leave this as "sdl2" and simply run with features=glfw.

File: Cargo.toml [hunk 2]
```diff
@@ -7,17 +7,18 @@ build = "build.rs"
...
+glfw = ["skulpin/skulpin_glfw"]
 winit = ["skulpin/skulpin_winit"]
 
 [dependencies]
 euclid = "0.20.7"
 font-kit = "0.6.0"
 skribo = { git = "https://github.com/linebender/skribo" }
 lru = "0.4.3"
-skulpin = { git = "https://github.com/aclysma/skulpin" }
+skulpin = { git = "https://github.com/uhuntu/skulpin", branch = "glfw-windows" }
```
Comment (j4qfrost) [inline Cargo.toml:21]: Get this validated with upstream first.

File: src/bridge/layouts/mod.rs [hunk 3]
```diff
@@ -123,6 +132,36 @@ pub fn produce_neovim_keybinding_string(
...
+    keycode: Option<Keycode>,
+    keytext: Option<String>,
+    modifiers: Option<Mod>,
+) -> Option<String> {
+    let mut shift = false;
+    let mut ctrl = false;
+    let mut alt = false;
+    let mut gui = false;
+    if let Some(modifiers) = modifiers {
+        shift = modifiers.contains(Mod::Shift);
```
Comment (j4qfrost) [inline src/bridge/layouts/mod.rs:146]: This is usually handled using bit shifts, so there isn't a need for an entire vector to handle the state. Is that how the modifier is implemented?

File: src/bridge/ui_commands.rs [hunk 4]
```diff
@@ -3,7 +3,9 @@ use nvim_rs::compat::tokio::Compat;
 use nvim_rs::Neovim;
 use tokio::process::ChildStdin;
 
+#[cfg(not(feature = "glfw"))]
```
Comment (j4qfrost) [inline src/bridge/ui_commands.rs:6]: This could possibly wait for #347 to be merged in.

File: src/bridge/ui_commands.rs [hunk 5]
```diff
@@ -56,6 +65,7 @@ impl UiCommand {
                         .expect("Mouse Input Failed");
                 }
             }
+            #[cfg(not(feature = "glfw"))]
```
Comment (j4qfrost) [inline src/bridge/ui_commands.rs:68]: This is pretty important to not include.

**Comment by j4qfrost** [comment]
Good, start but it's still got a ways away. I would start by getting your skulpin branch approved upstream first.

File: Cargo.toml (hunk 2 above)
Comment (uhuntu) [reply Cargo.toml:21]: https://github.com/aclysma/skulpin/pull/72

File: src/bridge/layouts/mod.rs (hunk 3 above)
Comment (uhuntu) [reply src/bridge/layouts/mod.rs:146]: Actually I did not completely get what you mean ? I just reference the code from both sdl2 and winit

File: src/bridge/ui_commands.rs (hunk 5 above)
Comment (uhuntu) [reply src/bridge/ui_commands.rs:68]: I know, for now, the branch is mostly prove the theory, I will try to implement the rest of them in spare time.

File: src/bridge/layouts/mod.rs (hunk 3 above)
Comment (j4qfrost) [reply src/bridge/layouts/mod.rs:146]: I double checked. This is fine. I thought this was self-implemented.

**Comment by j4qfrost** [comment]
I gave aclysma a heads up for your changes for glfw. Get that merged and implement scrolling, then I will run through it and do a final check.

File: src/bridge/ui_commands.rs (hunk 5 above)
Comment (uhuntu) [reply src/bridge/ui_commands.rs]: https://github.com/Kethku/neovide/pull/364/commits/a97fb234ea7eeefc5c8f0f3b89f54b8f227d6da4

File: src/bridge/ui_commands.rs (hunk 4 above)
Comment (uhuntu) [reply src/bridge/ui_commands.rs]: https://github.com/Kethku/neovide/pull/364/commits/a97fb234ea7eeefc5c8f0f3b89f54b8f227d6da4

File: src/window/glfw.rs [hunk 6]
```diff
@@ -273,7 +273,7 @@ pub fn initialize_settings() {
         .contains(&String::from("--noIdle"));
 
     SETTINGS.set(&WindowSettings {
-        refresh_rate: 60,
+        refresh_rate: 120,
```
Comment (j4qfrost) [inline src/window/glfw.rs:276]: Do you have a reason for changing this? Not trying to micromanage, just curious.

Comment (uhuntu) [reply src/window/glfw.rs:276]: Yes, you find the root point, I realized that glfw seems performance lower than sdl2, so I just increase that to 120, then seems performance improved. And I am not sure if it has any side effect, stay tuned.

Comment (Kethku) [reply src/window/glfw.rs:276]: Interesting. I will dig into this more once a full PR is ready. I suspect there  are minor differences in the way each windowing system queues a new frame.

File: src/bridge/ui_commands.rs (hunk 5 above)
Comment (uhuntu) [reply src/bridge/ui_commands.rs]: https://github.com/Kethku/neovide/pull/364/commits/2c3cde00d3fe4c9d54d8ef75a6afe0cad154f7c8

File: Cargo.toml (hunk 2 above)
Comment (uhuntu) [reply Cargo.toml]: https://github.com/uhuntu/neovide/commit/5a924dcfbc26b1d413ce52b1ce1495032b89b6a4

Comment (uhuntu) [reply Cargo.toml]: https://github.com/aclysma/skulpin/issues/68#issuecomment-699685177

**Comment by aclysma** [comment]
Regarding upstreaming GLFW support in skulpin, I left some comments here: https://github.com/aclysma/skulpin/issues/68#issuecomment-699685177

I'm happy to accept a PR for a GLFW implementation as long as it includes an example that works reasonably well on windows/mac/linux. I prototyped an implementation and example but it does not meet that requirement (on macOS anyways) as resizing the window makes the window contents permanently black.

While I don't want to upstream that implementation/example in it's broken state, I did upstream the changes that would be necessary to continue working on it out-of-tree. It's really quite small and it's very possible the bug is just bad example code!
 * Implementation: https://github.com/aclysma/skulpin/blob/6e54523f230d86f57c8a96612173aba584a343fd/skulpin-renderer-glfw/src/lib.rs
 * Example code: https://github.com/aclysma/skulpin/blob/6e54523f230d86f57c8a96612173aba584a343fd/examples/glfw_renderer_only.rs
 * It's in this branch: https://github.com/aclysma/skulpin/tree/glfw

**Comment by uhuntu** [comment]
As winit should be final choice, I would like to close this PR for this moment.

**Pull request closed without merging**